File: soohyun/python/programmers/0426/모두0으로만들기/ref.java
import java.util.ArrayList;

class Solution {
	static ArrayList<Integer>[] graph;
	static boolean[] visited;
	static long[] check;
	static long count=0;
	static long answer=0;
    public static long solution(int[] a, int[][] edges) {
    	
        graph = new ArrayList[a.length];
        visited = new boolean[a.length];
        check = new long[a.length];
        for(int i=0; i<a.length; i++)
        	graph[i] = new ArrayList<>();
        //입력 배열을 기반으로 트리 구성
        for(int i=0; i<edges.length; i++)
        {
        	int from = edges[i][0];
        	int to = edges[i][1];
        	graph[from].add(to);
        	graph[to].add(from);
        }
        visited[0] = true;
        dfs(0,a);
        
        
        System.out.println(answer);
        return answer;
    }
    public static void dfs(int cur,int[] a)
    {
    	for(int i=0; i<graph[cur].size(); i++)
    	{
    		if(visited[graph[cur].get(i)]==false)
    		{
    			visited[graph[cur].get(i)]=true;
    			dfs(graph[cur].get(i),a);
                //각 노드의 가중치 값을 더해서 올라오는 과정
    			check[cur] = check[cur]+check[graph[cur].get(i)];
                //가중치 연산 수행 횟수 계산
    			answer = answer+Math.abs(check[graph[cur].get(i)]);

    				
    		}
    	}
    	check[cur] = check[cur]+a[cur];
    	if(cur==0)
    	{
    		if(check[cur]!=0)
    			answer=-1;
    	}
    }
}

File: soohyun/python/programmers/0426/모두0으로만들기/ref.py

"""Solution code for "Programmers 76503. 모두 0으로 만들기

- Problem link: https://programmers.co.kr/learn/courses/30/lessons/76503
- Solution link: http://www.teferi.net/ps/problems/programmers/76503
"""

import sys

sys.setrecursionlimit(10**9)


def solution(a, edges):
    def dfs(u, parent=None):
        answer = 0
        for child in graph[u]: # 리프노드까지 이동
            if child != parent:
                answer += dfs(child, u)
        if parent is None: # 부모노드까지 온경우
            if a[u] != 0: # 부모노드가 0이 아닌경우
                return -1 
        else: # 부모노드가 아닌경우
            answer += abs(a[u]) # 답은 노드 값을 돌려줌
            a[parent] += a[u] # 부모값도 현재값을 더함
            print(answer, a)
        return answer

    graph = [[] for _ in a]
    for u, v in edges:
        graph[u].append(v)
        graph[v].append(u)
    return dfs(0)
#a	edges	result


solution([-5,0,2,1,2], [[0,1],[3,4],[2,3],[0,3]])
solution([0,1,0],[[0,1],[1,2]])